# Filtrado 18
Consiste en sacar todas las dwas asociadas por cada profesion con todo su work context. Paso 2: clasificacion del degree of automation en funcion del work context.

**FUTURA VERSION**

## SQL

In [1]:
# -*- coding: utf-8 -*-
from pathlib import Path

# Lista tal como la diste
ocupations = [
    'Industrial Production Managers','Quality Control Systems Managers','Supply Chain Managers',
    'Human Resources Managers','Logistics Engineers','Bioengineers and Biomedical Engineers',
    'Chemical Engineers','Validation Engineers','Manufacturing Engineers','Mechanical Engineers',
    'Automotive Engineers','Mechatronics Engineers','Robotics Engineers','Robotics Technicians',
    'Industrial Engineering Technologists and Technicians',
    'Mechanical Engineering Technologists and Technicians',
    'Biochemists and Biophysicists','Microbiologists','Chemists'
]


In [2]:

# Carpeta de salida (cámbiala si quieres)
OUT_DIR = Path("filtrado18")
OUT_DIR.mkdir(parents=True, exist_ok=True)

TEMPLATE = """USE onet;

SELECT
    o.onetsoc_code,
    o.title,
    td.dwa_id,
    dwaref.dwa_title,

    MAX(CASE WHEN cmr.element_name = 'Consequence of Error' THEN w.data_value END) AS consequence_of_error,
    MAX(CASE WHEN cmr.element_name = 'Contact With Others' THEN w.data_value END) AS contact_with_others,
    MAX(CASE WHEN cmr.element_name = 'Coordinate or Lead Others' THEN w.data_value END) AS coordinate_or_lead_others,
    MAX(CASE WHEN cmr.element_name = 'Cramped Work Space, Awkward Positions' THEN w.data_value END) AS cramped_work_space,
    MAX(CASE WHEN cmr.element_name = 'Deal With External Customers' THEN w.data_value END) AS deal_with_external_customers,
    MAX(CASE WHEN cmr.element_name = 'Deal With Physically Aggressive People' THEN w.data_value END) AS deal_with_physically_aggressive_people,
    MAX(CASE WHEN cmr.element_name = 'Deal With Unpleasant or Angry People' THEN w.data_value END) AS deal_with_unpleasant_or_angry_people,
    MAX(CASE WHEN cmr.element_name = 'Degree of Automation' THEN w.data_value END) AS degree_of_automation,
    MAX(CASE WHEN cmr.element_name = 'Electronic Mail' THEN w.data_value END) AS electronic_mail,
    MAX(CASE WHEN cmr.element_name = 'Exposed to Contaminants' THEN w.data_value END) AS exposed_to_contaminants,
    MAX(CASE WHEN cmr.element_name = 'Exposed to Disease or Infections' THEN w.data_value END) AS exposed_to_disease_or_infections,
    MAX(CASE WHEN cmr.element_name = 'Exposed to Hazardous Conditions' THEN w.data_value END) AS exposed_to_hazardous_conditions,
    MAX(CASE WHEN cmr.element_name = 'Exposed to Hazardous Equipment' THEN w.data_value END) AS exposed_to_hazardous_equipment,
    MAX(CASE WHEN cmr.element_name = 'Exposed to High Places' THEN w.data_value END) AS exposed_to_high_places,
    MAX(CASE WHEN cmr.element_name = 'Exposed to Minor Burns, Cuts, Bites, or Stings' THEN w.data_value END) AS exposed_to_minor_burns_cuts_bites_stings,
    MAX(CASE WHEN cmr.element_name = 'Exposed to Radiation' THEN w.data_value END) AS exposed_to_radiation,
    MAX(CASE WHEN cmr.element_name = 'Exposed to Whole Body Vibration' THEN w.data_value END) AS exposed_to_whole_body_vibration,
    MAX(CASE WHEN cmr.element_name = 'Extremely Bright or Inadequate Lighting' THEN w.data_value END) AS bright_or_inadequate_lighting,
    MAX(CASE WHEN cmr.element_name = 'Face-to-Face Discussions' THEN w.data_value END) AS face_to_face_discussions,
    MAX(CASE WHEN cmr.element_name = 'Freedom to Make Decisions' THEN w.data_value END) AS freedom_to_make_decisions,
    MAX(CASE WHEN cmr.element_name = 'Frequency of Conflict Situations' THEN w.data_value END) AS frequency_of_conflict_situations,
    MAX(CASE WHEN cmr.element_name = 'Impact of Decisions on Co-workers or Company Results' THEN w.data_value END) AS impact_of_decisions,
    MAX(CASE WHEN cmr.element_name = 'Importance of Being Exact or Accurate' THEN w.data_value END) AS importance_of_being_exact_or_accurate,
    MAX(CASE WHEN cmr.element_name = 'Importance of Repeating Same Tasks' THEN w.data_value END) AS importance_of_repeating_same_tasks,
    MAX(CASE WHEN cmr.element_name = 'In an Enclosed Vehicle or Equipment' THEN w.data_value END) AS in_enclosed_vehicle_or_equipment,
    MAX(CASE WHEN cmr.element_name = 'In an Open Vehicle or Equipment' THEN w.data_value END) AS in_open_vehicle_or_equipment,
    MAX(CASE WHEN cmr.element_name = 'Indoors, Environmentally Controlled' THEN w.data_value END) AS indoors_environmentally_controlled,
    MAX(CASE WHEN cmr.element_name = 'Indoors, Not Environmentally Controlled' THEN w.data_value END) AS indoors_not_environmentally_controlled,
    MAX(CASE WHEN cmr.element_name = 'Letters and Memos' THEN w.data_value END) AS letters_and_memos,
    MAX(CASE WHEN cmr.element_name = 'Level of Competition' THEN w.data_value END) AS level_of_competition,
    MAX(CASE WHEN cmr.element_name = 'Outdoors, Exposed to Weather' THEN w.data_value END) AS outdoors_exposed_to_weather,
    MAX(CASE WHEN cmr.element_name = 'Outdoors, Under Cover' THEN w.data_value END) AS outdoors_under_cover,
    MAX(CASE WHEN cmr.element_name = 'Pace Determined by Speed of Equipment' THEN w.data_value END) AS pace_determined_by_equipment,
    MAX(CASE WHEN cmr.element_name = 'Physical Proximity' THEN w.data_value END) AS physical_proximity,
    MAX(CASE WHEN cmr.element_name = 'Public Speaking' THEN w.data_value END) AS public_speaking,
    MAX(CASE WHEN cmr.element_name = 'Responsibility for Outcomes and Results' THEN w.data_value END) AS responsibility_for_outcomes,
    MAX(CASE WHEN cmr.element_name = 'Responsible for Others'' Health and Safety' THEN w.data_value END) AS responsible_for_others_health_and_safety,
    MAX(CASE WHEN cmr.element_name = 'Sounds, Noise Levels Are Distracting or Uncomfortable' THEN w.data_value END) AS noise_levels_distracting,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Bending or Twisting the Body' THEN w.data_value END) AS spend_time_bending_or_twisting,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Climbing Ladders, Scaffolds, or Poles' THEN w.data_value END) AS spend_time_climbing_ladders,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Keeping or Regaining Balance' THEN w.data_value END) AS spend_time_keeping_balance,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Kneeling, Crouching, Stooping, or Crawling' THEN w.data_value END) AS spend_time_kneeling_crouching,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Sitting' THEN w.data_value END) AS spend_time_sitting,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Standing' THEN w.data_value END) AS spend_time_standing,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Using Your Hands to Handle, Control, or Feel Objects, Tools, or Controls' THEN w.data_value END) AS spend_time_using_hands,
    MAX(CASE WHEN cmr.element_name = 'Spend Time Walking and Running' THEN w.data_value END) AS spend_time_walking_running,
    MAX(CASE WHEN cmr.element_name = 'Structured versus Unstructured Work' THEN w.data_value END) AS structured_vs_unstructured_work,
    MAX(CASE WHEN cmr.element_name = 'Telephone' THEN w.data_value END) AS telephone,
    MAX(CASE WHEN cmr.element_name = 'Very Hot or Cold Temperatures' THEN w.data_value END) AS very_hot_or_cold_temperatures,
    MAX(CASE WHEN cmr.element_name = 'Wear Common Protective or Safety Equipment such as Safety Shoes, Glasses, Gloves, Hearing Protection, Hard Hats, or Life Jackets' THEN w.data_value END) AS wear_common_protective_equipment,
    MAX(CASE WHEN cmr.element_name = 'Wear Specialized Protective or Safety Equipment such as Breathing Apparatus, Safety Harness, Full Protection Suits, or Radiation Protection' THEN w.data_value END) AS wear_specialized_protective_equipment,
    MAX(CASE WHEN cmr.element_name = 'Work With Work Group or Team' THEN w.data_value END) AS work_with_team
FROM tasks_to_dwas td
JOIN occupation_data o           ON td.onetsoc_code = o.onetsoc_code
JOIN dwa_reference dwaref        ON td.dwa_id      = dwaref.dwa_id
JOIN work_context w              ON o.onetsoc_code = w.onetsoc_code
JOIN content_model_reference cmr ON w.element_id   = cmr.element_id
WHERE o.title   = '{title_escaped}'
  AND w.scale_id = 'CX'
  AND cmr.element_name IN (
    'Consequence of Error','Contact With Others','Coordinate or Lead Others',
    'Cramped Work Space, Awkward Positions','Deal With External Customers',
    'Deal With Physically Aggressive People','Deal With Unpleasant or Angry People',
    'Degree of Automation','Electronic Mail','Exposed to Contaminants',
    'Exposed to Disease or Infections','Exposed to Hazardous Conditions',
    'Exposed to Hazardous Equipment','Exposed to High Places',
    'Exposed to Minor Burns, Cuts, Bites, or Stings','Exposed to Radiation',
    'Exposed to Whole Body Vibration','Extremely Bright or Inadequate Lighting',
    'Face-to-Face Discussions','Freedom to Make Decisions',
    'Frequency of Conflict Situations','Impact of Decisions on Co-workers or Company Results',
    'Importance of Being Exact or Accurate','Importance of Repeating Same Tasks',
    'In an Enclosed Vehicle or Equipment','In an Open Vehicle or Equipment',
    'Indoors, Environmentally Controlled','Indoors, Not Environmentally Controlled',
    'Letters and Memos','Level of Competition','Outdoors, Exposed to Weather',
    'Outdoors, Under Cover','Pace Determined by Speed of Equipment','Physical Proximity',
    'Public Speaking','Responsibility for Outcomes and Results',
    'Responsible for Others'' Health and Safety','Sounds, Noise Levels Are Distracting or Uncomfortable',
    'Spend Time Bending or Twisting the Body','Spend Time Climbing Ladders, Scaffolds, or Poles',
    'Spend Time Keeping or Regaining Balance','Spend Time Kneeling, Crouching, Stooping, or Crawling',
    'Spend Time Sitting','Spend Time Standing',
    'Spend Time Using Your Hands to Handle, Control, or Feel Objects, Tools, or Controls',
    'Spend Time Walking and Running','Structured versus Unstructured Work','Telephone',
    'Very Hot or Cold Temperatures',
    'Wear Common Protective or Safety Equipment such as Safety Shoes, Glasses, Gloves, Hearing Protection, Hard Hats, or Life Jackets',
    'Wear Specialized Protective or Safety Equipment such as Breathing Apparatus, Safety Harness, Full Protection Suits, or Radiation Protection',
    'Work With Work Group or Team'
  )
GROUP BY o.onetsoc_code, o.title, td.dwa_id, dwaref.dwa_title
ORDER BY dwaref.dwa_title;


"""


In [3]:

for title in ocupations:
    # title=title.lower()
    # Escapar comillas simples para SQL (por si en el futuro aparece alguna)
    title_escaped = title.replace("'", "''")
    # Nombre de archivo: espacios -> _
    filename = title.replace(" ", "_") + ".sql"
    sql_text = TEMPLATE.format(title_escaped=title_escaped)
    (OUT_DIR / filename).write_text(sql_text, encoding="utf-8")

print(f"Generados {len(ocupations)} archivos .sql en: {OUT_DIR.resolve()}")


Generados 19 archivos .sql en: /Users/jpardo/Desktop/Proyectos/AI4LABOUR/Mejoras/eerr/filtrado18


## Agrupacion de todos los datasets

In [4]:
import os
import pandas as pd

path = "/Users/jpardo/Desktop/Proyectos/AI4LABOUR/mejoras/eerr/filtrado18"

# Listar solo CSV
archivos = [i for i in os.listdir(path) if i.endswith('.csv')]

dfs = []  # lista para acumular dataframes

for i in archivos:
    path_archivo = os.path.join(path, i)
    df_temp = pd.read_csv(path_archivo)
    df_temp["archivo_origen"] = i  # opcional: para saber de qué archivo viene cada fila
    dfs.append(df_temp)

# Unir todos los DataFrames
df_final = pd.concat(dfs, ignore_index=True)

print(f"Se han cargado {len(df_final)} filas de {len(archivos)} archivos.")


Se han cargado 422 filas de 19 archivos.


In [5]:
display(df_final)

,onetsoc_code,title,dwa_id,dwa_title,consequence_of_error,contact_with_others,coordinate_or_lead_others,cramped_work_space,deal_with_external_customers,deal_with_physically_aggressive_people,...,spend_time_standing,spend_time_using_hands,spend_time_walking_running,structured_vs_unstructured_work,telephone,very_hot_or_cold_temperatures,wear_common_protective_equipment,wear_specialized_protective_equipment,work_with_team,archivo_origen
0,17-2141.02,Automotive Engineers,4.A.3.a.2.I09.D04,Calibrate scientific or technical equipment.,2.90,3.86,3.80,1.85,3.14,1.45,...,2.45,2.33,2.21,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv
1,17-2141.02,Automotive Engineers,4.A.2.a.4.I12.D01,Conduct quantitative failure analyses of opera...,2.90,3.86,3.80,1.85,3.14,1.45,...,2.45,2.33,2.21,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv
2,17-2141.02,Automotive Engineers,4.A.4.a.2.I11.D07,"Coordinate activities with suppliers, contract...",2.90,3.86,3.80,1.85,3.14,1.45,...,2.45,2.33,2.21,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv
3,17-2141.02,Automotive Engineers,4.A.2.b.2.I21.D01,Create models of engineering designs or methods.,2.90,3.86,3.80,1.85,3.14,1.45,...,2.45,2.33,2.21,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv
4,17-2141.02,Automotive Engineers,4.A.2.b.2.I14.D06,Design control systems for mechanical or other...,2.90,3.86,3.80,1.85,3.14,1.45,...,2.45,2.33,2.21,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
417,11-3051.01,Quality Control Systems Managers,4.A.3.b.6.I15.D06,Prepare operational progress or status reports.,2.76,4.83,4.45,1.27,3.57,1.42,...,2.67,1.78,2.31,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv
418,11-3051.01,Quality Control Systems Managers,4.A.4.b.6.I05.D07,Recommend organizational process or policy cha...,2.76,4.83,4.45,1.27,3.57,1.42,...,2.67,1.78,2.31,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv
419,11-3051.01,Quality Control Systems Managers,4.A.1.a.1.I02.D15,Review details of technical drawings or specif...,2.76,4.83,4.45,1.27,3.57,1.42,...,2.67,1.78,2.31,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv
420,11-3051.01,Quality Control Systems Managers,4.A.2.a.3.I01.D04,Review documents or materials for compliance w...,2.76,4.83,4.45,1.27,3.57,1.42,...,2.67,1.78,2.31,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv


In [6]:
!pip install openpyxl


[notice] A new release of pip is available: 25.1.1 -> 25.2
[notice] To update, run: pip install --upgrade pip


In [7]:
# Cargar el Excel
ruta_excel = "/Users/jpardo/Desktop/Proyectos/AI4LABOUR/mejoras/eerr/Jobs List ICBE.xlsx"
df_excel = pd.read_excel(ruta_excel, sheet_name="ICBE")
# Pasar nombres de columnas a minúsculas
df_excel.columns = df_excel.columns.str.lower()


display(df_excel)

,o.net.soc.code,title,icbe,description
0,11-3051.00,Industrial Production Managers,Primary,"Plan, direct, or coordinate the work activitie..."
1,11-3051.01,Quality Control Systems Managers,Secondary,"Plan, direct, or coordinate quality assurance ..."
2,11-3071.04,Supply Chain Managers,Secondary,"Direct or coordinate production, purchasing, w..."
3,11-3121.00,Human Resources Managers,Secondary,"Plan, direct, or coordinate human resources ac..."
4,13-1081.01,Logistics Engineers,Secondary,Design or analyze operational solutions for pr...
5,17-2031.00,Bioengineers and Biomedical Engineers,NaN,"Apply knowledge of engineering, biology, chemi..."
6,17-2041.00,Chemical Engineers,Primary,Design chemical plant equipment and devise pro...
7,17-2112.02,Validation Engineers,NaN,Design or plan protocols for equipment or proc...
8,17-2112.03,Manufacturing Engineers,Primary,"Design, integrate, or improve manufacturing sy..."
9,17-2141.00,Mechanical Engineers,Primary,Perform engineering duties in planning and des...


In [8]:
# Unir DataFrames
# Aquí debes definir la columna clave común en ambos DF, por ejemplo 'dwa_id'
df_merged = pd.merge(df_final, df_excel, on="title", how="left")

print(f"DataFrame final tiene {df_merged.shape[0]} filas y {df_merged.shape[1]} columnas.")

DataFrame final tiene 422 filas y 60 columnas.


In [9]:
display(df_merged)

,onetsoc_code,title,dwa_id,dwa_title,consequence_of_error,contact_with_others,coordinate_or_lead_others,cramped_work_space,deal_with_external_customers,deal_with_physically_aggressive_people,...,structured_vs_unstructured_work,telephone,very_hot_or_cold_temperatures,wear_common_protective_equipment,wear_specialized_protective_equipment,work_with_team,archivo_origen,o.net.soc.code,icbe,description
0,17-2141.02,Automotive Engineers,4.A.3.a.2.I09.D04,Calibrate scientific or technical equipment.,2.90,3.86,3.80,1.85,3.14,1.45,...,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv,17-2141.02,NaN,Develop new or improved designs for vehicle st...
1,17-2141.02,Automotive Engineers,4.A.2.a.4.I12.D01,Conduct quantitative failure analyses of opera...,2.90,3.86,3.80,1.85,3.14,1.45,...,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv,17-2141.02,NaN,Develop new or improved designs for vehicle st...
2,17-2141.02,Automotive Engineers,4.A.4.a.2.I11.D07,"Coordinate activities with suppliers, contract...",2.90,3.86,3.80,1.85,3.14,1.45,...,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv,17-2141.02,NaN,Develop new or improved designs for vehicle st...
3,17-2141.02,Automotive Engineers,4.A.2.b.2.I21.D01,Create models of engineering designs or methods.,2.90,3.86,3.80,1.85,3.14,1.45,...,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv,17-2141.02,NaN,Develop new or improved designs for vehicle st...
4,17-2141.02,Automotive Engineers,4.A.2.b.2.I14.D06,Design control systems for mechanical or other...,2.90,3.86,3.80,1.85,3.14,1.45,...,4.19,4.65,2.10,3.24,1.95,4.65,Automotive_Engineers.csv,17-2141.02,NaN,Develop new or improved designs for vehicle st...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
417,11-3051.01,Quality Control Systems Managers,4.A.3.b.6.I15.D06,Prepare operational progress or status reports.,2.76,4.83,4.45,1.27,3.57,1.42,...,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv,11-3051.01,Secondary,"Plan, direct, or coordinate quality assurance ..."
418,11-3051.01,Quality Control Systems Managers,4.A.4.b.6.I05.D07,Recommend organizational process or policy cha...,2.76,4.83,4.45,1.27,3.57,1.42,...,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv,11-3051.01,Secondary,"Plan, direct, or coordinate quality assurance ..."
419,11-3051.01,Quality Control Systems Managers,4.A.1.a.1.I02.D15,Review details of technical drawings or specif...,2.76,4.83,4.45,1.27,3.57,1.42,...,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv,11-3051.01,Secondary,"Plan, direct, or coordinate quality assurance ..."
420,11-3051.01,Quality Control Systems Managers,4.A.2.a.3.I01.D04,Review documents or materials for compliance w...,2.76,4.83,4.45,1.27,3.57,1.42,...,4.49,4.62,2.09,3.43,1.35,4.80,Quality_Control_Systems_Managers.csv,11-3051.01,Secondary,"Plan, direct, or coordinate quality assurance ..."


# Ocuppations

In [10]:
from pprint import pprint
occupations = df_excel['title'].tolist()
pprint(occupations)

['Industrial Production Managers',
 'Quality Control Systems Managers',
 'Supply Chain Managers',
 'Human Resources Managers',
 'Logistics Engineers',
 'Bioengineers and Biomedical Engineers',
 'Chemical Engineers',
 'Validation Engineers',
 'Manufacturing Engineers',
 'Mechanical Engineers',
 'Automotive Engineers',
 'Mechatronics Engineers',
 'Robotics Engineers',
 'Robotics Technicians',
 'Industrial Engineering Technologists and Technicians',
 'Mechanical Engineering Technologists and Technicians',
 'Biochemists and Biophysicists',
 'Microbiologists',
 'Chemists']


In [11]:
print(len(occupations))

19


# DWAs

In [12]:
conteo = pd.DataFrame(df_final["dwa_title"].value_counts())

display(conteo)


,count
dwa_title,
Estimate operational costs.,8
Design industrial processing systems.,7
"Recommend technical design or process changes to improve efficiency, quality, or performance.",7
Maintain operational records or records systems.,6
Train personnel on proper operational procedures.,6
...,...
Coordinate with external parties to exchange information.,1
Repair electronic equipment.,1
"Maintain inventories of materials, equipment, or products.",1


In [13]:
num_compartidos = conteo[conteo['count'] > 1]
display(num_compartidos)

,count
dwa_title,
Estimate operational costs.,8
Design industrial processing systems.,7
"Recommend technical design or process changes to improve efficiency, quality, or performance.",7
Maintain operational records or records systems.,6
Train personnel on proper operational procedures.,6
...,...
"Investigate system, equipment, or product failures.",2
Research diseases or parasites.,2
Design electronic or computer equipment or instrumentation.,2


In [14]:
221-100

121

Hay 121 dwa que se comparten entre las 19 occupations. 

## Filtrado 18 - mean

In [15]:
# Asegurar que las columnas numéricas sean numéricas
for col in df_final.columns:
    if col not in ["dwa_title", "title"]:
        df_final[col] = pd.to_numeric(df_final[col], errors="coerce")

# Agrupar por DWA y calcular medias
degree_avg_dwa = (
    df_final
    .groupby("dwa_title")
    .mean(numeric_only=True)
    .reset_index()
)

display(degree_avg_dwa)


,dwa_title,onetsoc_code,dwa_id,consequence_of_error,contact_with_others,coordinate_or_lead_others,cramped_work_space,deal_with_external_customers,deal_with_physically_aggressive_people,deal_with_unpleasant_or_angry_people,...,spend_time_standing,spend_time_using_hands,spend_time_walking_running,structured_vs_unstructured_work,telephone,very_hot_or_cold_temperatures,wear_common_protective_equipment,wear_specialized_protective_equipment,work_with_team,archivo_origen
0,Administer compensation or benefits programs.,NaN,NaN,3.420000,4.620,3.880000,1.080000,2.880000,2.000000,3.290000,...,2.0400,2.540000,1.8800,4.710000,5.000000,1.210000,1.460,1.000000,4.250000,NaN
1,Administer standardized physical or psychologi...,NaN,NaN,3.420000,4.620,3.880000,1.080000,2.880000,2.000000,3.290000,...,2.0400,2.540000,1.8800,4.710000,5.000000,1.210000,1.460,1.000000,4.250000,NaN
2,Advise customers on technical or procedural is...,NaN,NaN,2.760000,4.830,4.450000,1.270000,3.570000,1.420000,2.990000,...,2.6700,1.780000,2.3100,4.490000,4.620000,2.090000,3.430,1.350000,4.800000,NaN
3,Advise customers on the use of products or ser...,NaN,NaN,3.223333,4.150,3.556667,1.783333,2.906667,1.333333,2.486667,...,2.5100,2.713333,2.0200,4.246667,4.530000,1.910000,3.660,1.740000,4.393333,NaN
4,Advise others on career or personal development.,NaN,NaN,3.420000,4.620,3.880000,1.080000,2.880000,2.000000,3.290000,...,2.0400,2.540000,1.8800,4.710000,5.000000,1.210000,1.460,1.000000,4.250000,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
216,Test products for functionality or quality.,NaN,NaN,3.570000,4.535,3.865000,2.585000,2.960000,1.245000,2.745000,...,3.1800,3.175000,2.8850,4.025000,4.475000,2.420000,4.505,1.680000,4.265000,NaN
217,Test quality of materials or finished products.,NaN,NaN,3.420000,3.820,3.550000,1.200000,2.980000,1.250000,2.430000,...,2.9000,2.580000,2.1600,4.140000,3.820000,1.140000,4.760,1.700000,3.960000,NaN
218,Train personnel on proper operational procedures.,NaN,NaN,3.291667,4.090,3.638333,2.021667,2.836667,1.186667,2.455000,...,2.5400,2.760000,2.1800,4.128333,4.446667,1.943333,3.875,1.746667,4.381667,NaN
219,Update technical knowledge.,NaN,NaN,3.010000,4.055,3.660000,1.792500,2.885000,1.325000,2.530000,...,2.3575,2.342500,2.0275,4.080000,4.520000,1.905000,3.370,1.777500,4.480000,NaN


In [16]:
import re, unicodedata

# --- Regex precompiles ---
RE_ZERO_WIDTH   = re.compile(r'[\u200B-\u200F\u202A-\u202E\u2060\uFEFF]')
RE_SPACES       = re.compile(r'\s+')
RE_TRAILING_DOT = re.compile(r'[.\s]+$')

# Clase de comillas (ASCII + tipográficas + angulares)
QUOTE_UTF8_CHARS = '"\'`“”„‟‘’‚‛«»‹›'
RE_EDGE_QUOTES   = re.compile(rf'^[{re.escape(QUOTE_UTF8_CHARS)}\s]+|[{re.escape(QUOTE_UTF8_CHARS)}\s]+$')

# Mojibake -> UTF-8/ASCII
MOJIBAKE_MAP = {
    "•": "•",
    "—": "—",
    "–": "–",
    "…": "…",
    "’": "'",
    "‘": "'",
    "“": '"',
    "”": '"',
    "​": "",
}

# Enumeraciones iniciales (NO metas comillas aquí)
LEADING_ENUM_PATTERNS = [
    re.compile(r'^\s*\d+\s*\t+\s*'),               # "12\t..."
    re.compile(r'^\s*\d+\s*[\.\)]\s*'),            # "1." / "2)"
    re.compile(r'^\s*[\(\[]\s*\d+\s*[\)\]]\s*'),   # "(1)" / "[2]"
    re.compile(r'^\s*\d+\s*[-–—]\s+'),             # "1 - " / "1 – " / "1 — "
    re.compile(r'^\s*[A-Za-z]\s*[\.\)]\s+'),       # "a) " / "B. "
    re.compile(r'^\s*[ivxlcdmIVXLCDM]+\s*[\.\)]\s+'),  # "IV. " / "i) "
    re.compile(r'^\s*[•\-\*\u2219·–—]\s*\t+\s*'),  # bullets con tab
    re.compile(r'^\s*[•\-\*\u2219·–—]\s+'),        # bullets con espacio
]

def _strip_edge_quotes_loop(s: str) -> str:
    """Recorta comillas/espacios en bordes de forma repetida (izq./der.)."""
    while True:
        new_s = RE_EDGE_QUOTES.sub('', s).strip()
        if new_s == s:
            return s
        s = new_s

def clean_skill(text: str) -> str:
    """
    Devuelve la skill limpia en minúsculas:
      - quita enumeraciones iniciales,
      - recorta comillas de borde (ASCII/UTF-8),
      - corrige mojibake, elimina zero-width/bidi/BOM,
      - quita punto(s) finales,
      - conserva paréntesis internos.
    """
    if text is None:
        return ""
    s = unicodedata.normalize("NFKC", str(text))

    # Mojibake -> UTF8/ASCII
    for bad, good in MOJIBAKE_MAP.items():
        if bad in s:
            s = s.replace(bad, good)

    # Invisibles
    s = RE_ZERO_WIDTH.sub("", s)

    # Enumeraciones al inicio (pelar en capas)
    prev = None
    while s != prev:
        prev = s
        for pat in LEADING_ENUM_PATTERNS:
            s = pat.sub("", s, count=1)

    # Normaliza espacios
    s = RE_SPACES.sub(" ", s).strip()

    # 1ª pasada: recortar comillas de borde
    s = _strip_edge_quotes_loop(s)

    # Quitar punto(s) finales
    s = RE_TRAILING_DOT.sub("", s).strip()

    # 2ª pasada: recortar comillas de borde (para casos como ”.)
    s = _strip_edge_quotes_loop(s)

    # Minúsculas
    return s.lower().strip()

def _t(inp, expected):
    got = clean_skill(inp)
    assert got == expected, f"\nINPUT : {repr(inp)}\nGOT   : {repr(got)}\nEXPECT: {repr(expected)}"

# Comillas (ASCII, tipográficas, angulares) + punto final
_t('"Python Programming"', 'python programming')
_t('“Python Programming”.', 'python programming')
_t("‘Skill’", 'skill')
_t("«Skill»", 'skill')
_t("‹Skill›", 'skill')
_t("`Skill`", 'skill')
_t("' Skill '", 'skill')

# Punto final en skills largas
_t("Long descriptive skill.", "long descriptive skill")
_t("C++.", "c++")

# Mayúsculas/minúsculas
_t("JAVA", "java")
_t("Title Case Mixed", "title case mixed")

# Enumeraciones (número + tab / . / ) / (1) / [2] / 1 - / letra) / romanos / bullets
_t("12\tSystems Design", "systems design")
_t("1.\tSkill", "skill")
_t("(1)\tSkill", "skill")
_t("[2] Skill", "skill")
_t("1 - Skill", "skill")
_t("a)\tSkill", "skill")
_t("IV.\tSkill", "skill")
_t("•\tLeadership", "leadership")
_t("-\tProject Management", "project management")
_t("—\tSystems", "systems")
_t("–\tSystems", "systems")
_t("• bullet style skill", "bullet style skill")
_t("·\tAnother Skill", "another skill")

# Mojibake en medio del texto + zero-width
_t("4A’s Marketing", "4a's marketing")
_t("H​ow to Design", "how to design")

# Paréntesis preservados
_t("Design (CAD)", "design (cad)")
_t("A/B Testing (advanced)", "a/b testing (advanced)")


In [17]:
temp=degree_avg_dwa['dwa_title'].apply(clean_skill)


display(temp)

0           administer compensation or benefits programs
1      administer standardized physical or psychologi...
2      advise customers on technical or procedural is...
3      advise customers on the use of products or ser...
4        advise others on career or personal development
                             ...                        
216           test products for functionality or quality
217       test quality of materials or finished products
218     train personnel on proper operational procedures
219                           update technical knowledge
220                                write grant proposals
Name: dwa_title, Length: 221, dtype: object

In [18]:
conteo = pd.DataFrame(temp.value_counts())

display(conteo)

,count
dwa_title,
administer compensation or benefits programs,1
monitor environmental impacts of production or development activities,1
maintain laboratory or technical equipment,1
maintain operational records or records systems,1
maintain personnel records,1
...,...
develop specifications for new products or processes,1
develop sustainable business strategies or practices,1
develop sustainable organizational policies or practices,1


In [19]:
for i in temp:
    print(i) 


administer compensation or benefits programs
administer standardized physical or psychological tests
advise customers on technical or procedural issues
advise customers on the use of products or services
advise others on career or personal development
advise others on legal or regulatory compliance matters
advise others on logistics topics
advise others regarding green practices or environmental concerns
analyze biological samples
analyze chemical compounds or substances
analyze costs and benefits of proposed designs or projects
analyze data to assess operational or project effectiveness
analyze data to inform operational decisions or activities
analyze data to inform personnel decisions
analyze design or requirements information for mechanical equipment or systems
analyze environmental regulations to ensure organizational compliance
analyze green technology design requirements
analyze jobs using observation, survey, or interview techniques
analyze logistics processes
analyze operation

In [20]:
degree_avg_dwa.shape

(221, 56)

In [21]:
degree_avg_dwa2 = degree_avg_dwa.copy()
degree_avg_dwa2['dwa_title'] = temp

display(degree_avg_dwa2)

,dwa_title,onetsoc_code,dwa_id,consequence_of_error,contact_with_others,coordinate_or_lead_others,cramped_work_space,deal_with_external_customers,deal_with_physically_aggressive_people,deal_with_unpleasant_or_angry_people,...,spend_time_standing,spend_time_using_hands,spend_time_walking_running,structured_vs_unstructured_work,telephone,very_hot_or_cold_temperatures,wear_common_protective_equipment,wear_specialized_protective_equipment,work_with_team,archivo_origen
0,administer compensation or benefits programs,NaN,NaN,3.420000,4.620,3.880000,1.080000,2.880000,2.000000,3.290000,...,2.0400,2.540000,1.8800,4.710000,5.000000,1.210000,1.460,1.000000,4.250000,NaN
1,administer standardized physical or psychologi...,NaN,NaN,3.420000,4.620,3.880000,1.080000,2.880000,2.000000,3.290000,...,2.0400,2.540000,1.8800,4.710000,5.000000,1.210000,1.460,1.000000,4.250000,NaN
2,advise customers on technical or procedural is...,NaN,NaN,2.760000,4.830,4.450000,1.270000,3.570000,1.420000,2.990000,...,2.6700,1.780000,2.3100,4.490000,4.620000,2.090000,3.430,1.350000,4.800000,NaN
3,advise customers on the use of products or ser...,NaN,NaN,3.223333,4.150,3.556667,1.783333,2.906667,1.333333,2.486667,...,2.5100,2.713333,2.0200,4.246667,4.530000,1.910000,3.660,1.740000,4.393333,NaN
4,advise others on career or personal development,NaN,NaN,3.420000,4.620,3.880000,1.080000,2.880000,2.000000,3.290000,...,2.0400,2.540000,1.8800,4.710000,5.000000,1.210000,1.460,1.000000,4.250000,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
216,test products for functionality or quality,NaN,NaN,3.570000,4.535,3.865000,2.585000,2.960000,1.245000,2.745000,...,3.1800,3.175000,2.8850,4.025000,4.475000,2.420000,4.505,1.680000,4.265000,NaN
217,test quality of materials or finished products,NaN,NaN,3.420000,3.820,3.550000,1.200000,2.980000,1.250000,2.430000,...,2.9000,2.580000,2.1600,4.140000,3.820000,1.140000,4.760,1.700000,3.960000,NaN
218,train personnel on proper operational procedures,NaN,NaN,3.291667,4.090,3.638333,2.021667,2.836667,1.186667,2.455000,...,2.5400,2.760000,2.1800,4.128333,4.446667,1.943333,3.875,1.746667,4.381667,NaN
219,update technical knowledge,NaN,NaN,3.010000,4.055,3.660000,1.792500,2.885000,1.325000,2.530000,...,2.3575,2.342500,2.0275,4.080000,4.520000,1.905000,3.370,1.777500,4.480000,NaN


In [22]:
degree_avg_dwa2.columns


Index(['dwa_title', 'onetsoc_code', 'dwa_id', 'consequence_of_error',
       'contact_with_others', 'coordinate_or_lead_others',
       'cramped_work_space', 'deal_with_external_customers',
       'deal_with_physically_aggressive_people',
       'deal_with_unpleasant_or_angry_people', 'degree_of_automation',
       'electronic_mail', 'exposed_to_contaminants',
       'exposed_to_disease_or_infections', 'exposed_to_hazardous_conditions',
       'exposed_to_hazardous_equipment', 'exposed_to_high_places',
       'exposed_to_minor_burns_cuts_bites_stings', 'exposed_to_radiation',
       'exposed_to_whole_body_vibration', 'bright_or_inadequate_lighting',
       'face_to_face_discussions', 'freedom_to_make_decisions',
       'frequency_of_conflict_situations', 'impact_of_decisions',
       'importance_of_being_exact_or_accurate',
       'importance_of_repeating_same_tasks',
       'in_enclosed_vehicle_or_equipment', 'in_open_vehicle_or_equipment',
       'indoors_environmentally_controlled',

In [23]:
# Supongamos que tu DataFrame se llama df
cols_a_eliminar = ["onetsoc_code", "dwa_id"]

degree_avg_dwa2 = degree_avg_dwa2.drop(columns=cols_a_eliminar)

In [24]:
degree_avg_dwa2.shape

(221, 54)

In [25]:
import csv

formato =[".csv", ".tsv"] 

OUT_DIR2 = Path(path+"/results")
OUT_DIR2.mkdir(parents=True, exist_ok=True)

for i in formato:
    degree_avg_dwa2.to_csv(f"{OUT_DIR2}/filtrado18_results_num{i}", index=False, encoding="utf-8",
                        quoting=csv.QUOTE_MINIMAL, quotechar='"', escapechar='\\')
